src/Spago/Build.hs

```diff
@@ -376,25 +412,29 @@ runBackend maybeBackend RunDirectories{ sourceDir, executeDir } moduleName maybe
         ExitFailure n -> die [ display failureMessage <> "Backend " <> displayShow backend <> " exited with error:" <> repr n ]
 
 
-bundleWithEsbuild :: HasLogFunc env => WithMain -> ModuleName -> TargetPath -> Platform -> Minify -> RIO env ()
-bundleWithEsbuild withMain (ModuleName moduleName) (TargetPath targetPath) platform minify = do
+bundleWithEsbuild :: HasLogFunc env => WithMain -> WithSrcMap -> ModuleName -> TargetPath -> Platform -> Minify -> RIO env ()
+bundleWithEsbuild withMain srcMap (ModuleName moduleName) (TargetPath targetPath) platform minify = do
   esbuild <- getESBuild
   let
     platformOpt = case platform of
-      Browser -> "browser"
-      Node -> "node"
+      Browser -> " --platform=browser"
+      Node -> " --platform=node"
     minifyOpt = case minify of
       NoMinify -> ""
       Minify -> " --minify"
+    srcMapOpt = case srcMap of
+      WithSrcMap -> " --sourcemap"
+      WithoutSrcMap -> ""
+    esbuildBase = esbuild <> platformOpt <> minifyOpt <> srcMapOpt <> " --format=esm --bundle --outfile=" <> targetPath
     cmd = case withMain of
-      WithMain ->
-        "echo \"import { main } from './output/" <> moduleName <> "/index.js'\nmain()\" | "
-        <> esbuild <> " --platform=" <> platformOpt <> minifyOpt <> " --bundle "
-        <> " --outfile=" <> targetPath
+      WithMain -> do
+        let
+          echoLine = case SysInfo.os of
+            "mingw32" -> "echo import { main } from './output/" <> moduleName <> "/index.js'; main(); | "
```

Comment:
Why are we running `echo` instead of writing to a file? I thought this change was about running from a file again?

Reply:
Not quite.

When `spago run` or `spago test` is called, then we do need to write the wrapper to a file, so that we call `node run.js` rather than `node --eval "..."`. CLI programs will drop the first two args (path to Node executable, path to entry file) to get the CLI args to the program. When using a file (i.e. `node run.js`), the CLI program will get its args correctly. When using a script (i.e. `node --eval ...`), Node only prefixes the CLI args with one arg (path to Node executable) but not a second arg. Thus, the CLI program will drop its first arg, thinking its the path to the script. So, we have to use a file there.

The above situation is for a different command: `spago bundle-app`. We're trying to bundle the outputted JavaScript code via `esbuild`. However, let's say we run `esbuild --bundle output/Main/index.js --outfile=index.js` and then `node index.js`. When we run `index.js`, nothing will happen because the `main` function was never called (e.g. `main()`). Thus, the above wrapper code is needed to call `main`. Since we're bundling the echo line, which imports `output/Main/index.js`, the resulting `index.js` file will actually run instead of sitting there looking pretty.

The reason why we are using an echo line here rather than writing to a file directly is to avoid the potential 'space in file name' issue that caused problems for `spago run` and other related file system issues. However, the echo line is really just a bad way of passing in the desired `stdin` directly to the `esbuild` command via the Process API, which is what https://github.com/working-group-purescript-es/spago/pull/3 fixes.